apps/rush-lib/src/logic/TaskSelector.ts

```diff
@@ -67,52 +67,60 @@ export class TaskSelector {
   }
 
   private _registerToFlags(toFlags: ReadonlyArray<string>): void {
+    const dependencies: Set<string> = new Set<string>();
+
     for (const toFlag of toFlags) {
       const toProject: RushConfigurationProject | undefined =
         this._options.rushConfiguration.findProjectByShorthandName(toFlag);
       if (!toProject) {
         throw new Error(`The project '${toFlag}' does not exist in rush.json`);
       }
 
-      const deps: Set<string> = this._collectAllDependencies(toProject.packageName);
+      this._collectAllDependencies(toProject.packageName, dependencies);
+    }
 
-      // Register any dependencies it may have
-      deps.forEach(dep => this._registerTask(this._options.rushConfiguration.getProjectByName(dep)));
+    // Register any dependencies it may have
+    dependencies.forEach(dep => this._registerTask(this._options.rushConfiguration.getProjectByName(dep)));
 
-      if (!this._options.ignoreDependencyOrder) {
-        // Add ordering relationships for each dependency
-        deps.forEach(dep => this._taskCollection.addDependencies(dep, this._rushLinkJson.localLinks[dep] || []));
-      }
+    if (!this._options.ignoreDependencyOrder) {
+      // Add ordering relationships for each dependency
+      dependencies.forEach(dependency => {
+        this._taskCollection.addDependencies(
+          dependency,
+          this._rushLinkJson.localLinks[dependency] || []
+        );
+      });
     }
   }
 
   private _registerFromFlags(fromFlags: ReadonlyArray<string>): void {
+    this._buildDependentGraph();
+    const dependents: Set<string> = new Set<string>();
+
     for (const fromFlag of fromFlags) {
       const fromProject: RushConfigurationProject | undefined
         = this._options.rushConfiguration.findProjectByShorthandName(fromFlag);
       if (!fromProject) {
         throw new Error(`The project '${fromFlag}' does not exist in rush.json`);
       }
 
-      // Only register projects which depend on the current package, as well as things that depend on them
-      this._buildDependentGraph();
+      this._collectAllDependents(fromProject.packageName, dependents);
+    }
 
-      // We will assume this project will be built, but act like it has no dependencies
-      const dependents: Set<string> = this._collectAllDependents(fromProject.packageName);
-      dependents.add(fromProject.packageName);
+    // Register all downstream dependents
+    dependents.forEach(dependent => {
```

Comment:
Can you use `for (const dependent of dependents) {` instead of `forEach`?

`forEach` is inefficient, and in TS 3.5 I observed some compiler bugs when the callback modifies closure variables (due to the type narrowing not realizing that the variable was modified).